### PR TITLE
[OPPRO-319] Separate fetch code/compile/package for external projects in gluten

### PR DESCRIPTION
**Pull request by zhejiangxiaomai**
## What changes were proposed in this pull request?
1. remove unnecessary build option in maven.
2. use shell to compile arrow and velox directly.
3. use maven compile Java and package only.
4. add check_commit function in build_velox.sh, make sure local branch contains latest commit of OAP/main.

### Build arrow-backend(Gazelle) guide
```
## install gcc and libraries to build arrow
apt-get update && apt-get install -y sudo locales wget tar tzdata git ccache cmake ninja-build build-essential llvm-11-dev clang-11 libiberty-dev libdwarf-dev libre2-dev libz-dev libssl-dev libboost-all-dev libcurl4-openssl-dev openjdk-8-jdk maven

## make sure jdk8 is used
export JAVA_HOME=/usr/lib/jvm/java-8-openjdk-amd64
export PATH=$JAVA_HOME/bin:$PATH

#setup proxy if needed
export http_proxy=xxxx
export https_proxy=xxxx

##config maven proxy
mkdir ~/.m2/
apt install vim
vim ~/.m2/settings.xml

#clone gluten 
git clone https://github.com/oap-project/gluten.git

## fetch arrow for gazelle and compile
cd gluten/ep/build-arrow/src
## please set GLUTEN_DIR at get_arrow.sh
./get_arrow.sh --arrow_branch=arrow-8.0.0-gluten-20220427a --gluten_dir=/gluten_home
./build_arrow_for_gazelle.sh --gluten_dir=/gluten_home

## compile gluten cpp
cd gluten/cpp

## please set GLUTEN_DIR at compile.sh
./compile.sh --build_gazelle_cpp_backend=ON --build_protobuf=ON --gluten_dir=/gluten_home

## compile gluten jvm and package
cd /gluten
mvn clean package -Pspark-3.2 -Pbackends-gazelle -DskipTests
```

### Build velox-backend guide
```
## install gcc and libraries to build arrow
apt-get update && apt-get install -y sudo locales wget tar tzdata git ccache cmake ninja-build build-essential llvm-11-dev clang-11 libiberty-dev libdwarf-dev libre2-dev libz-dev libssl-dev libboost-all-dev libcurl4-openssl-dev openjdk-8-jdk maven

## make sure jdk8 is used
export JAVA_HOME=/usr/lib/jvm/java-8-openjdk-amd64
export PATH=$JAVA_HOME/bin:$PATH

#setup proxy if needed
export http_proxy=xxxx
export https_proxy=xxxx

##config maven proxy
mkdir ~/.m2/
apt install vim
vim ~/.m2/settings.xml

#clone gluten
git clone https://github.com/oap-project/gluten.git

## fetch arrow and compile
cd /gluten_path/ep/build-arrow/src

## please set GLUTEN_DIR in get_arrow.sh and build_arrow_for_velox.sh
./get_arrow.sh --gluten_dir=/gluten_home
./build_arrow_for_velox.sh --gluten_dir=/gluten_home

## fetch velox and compile
## please set GLUTEN_DIR at get_velox.sh and build_velox.sh
cd /gluten_path/ep/build-velox/src
./get_velox.sh --gluten_dir=/gluten_home

## install protobuf at first time. we can set it to OFF when we recompile gluten.
./build_velox.sh --build_protobuf=ON --build_folly=ON  --gluten_dir=/gluten_home

## compile gluten cpp
## please set GLUTEN_DIR at compile.sh
cd /gluten/cpp
./compile.sh --build_velox_backend=ON  --gluten_dir=/gluten_home

## compile gluten jvm and package
cd /gluten
mvn clean package -Pbackends-velox -Pspark-3.2
```

## How was this patch tested?
Build velox-backend and arrow-backend on clean ubuntu docker successfully. need fix doc and pass CI.

### CODE REVIEW

File: ep/build-arrow/src/build_arrow_for_velox.sh [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+NPROC=$(nproc --ignore=2)
+TESTS=OFF
+BENCHMARKS=OFF
+ENABLE_EP_CACHE=OFF
+BUILD_TYPE=RelWithDebInfo
+
+echo "CMAKE Arguments:"
+echo "TESTS=${TESTS}"
+echo "ARROW_ROOT=${ARROW_ROOT}"
```
Comment (zuochunwei) [inline ep/build-arrow/src/build_arrow_for_velox.sh:17]: useless

File: ep/build-arrow/src/build_arrow_for_velox.sh [hunk 2]
```diff
@@ -0,0 +1,84 @@
+#!/bin/bash
+
+set -exu
+
+#setting gluten root path
+GLUTEN_DIR="/gluten"
+ARROW_ROOT="$GLUTEN_DIR/ep/build-arrow/build/arrow_install"
```
Comment (zuochunwei) [inline ep/build-arrow/src/build_arrow_for_velox.sh:7]: delete useless variables, make script simple

File: ep/build-velox/src/build_velox.sh [hunk 3]
```diff
@@ -1,72 +1,23 @@
...
+set -exu
 
-NPROC=$(nproc)
+GLUTEN_DIR=/gluten
+VELOX_HOME="$GLUTEN_DIR/ep/build-velox/build/velox_ep/"
 
-BUILD_VELOX_FROM_SOURCE=OFF
-COMPILE_VELOX=OFF
-ENABLE_EP_CACHE=OFF
 ENABLE_S3=OFF
```
Comment (zuochunwei) [inline ep/build-velox/src/build_velox.sh:8]: ENABLE_S3、ENABLE_HDFS are unused, why not just delete them?

File: ep/build-arrow/src/build_arrow_for_velox.sh (hunk 1 above)
Comment (zhejiangxiaomai) [reply ep/build-arrow/src/build_arrow_for_velox.sh:17]: Thank you, I am going to delete after I finish debugging.

File: .github/workflows/unittests.yml [hunk 4]
```diff
@@ -27,38 +27,33 @@ concurrency:
 jobs:
 
   arrow-backend-test:
-    runs-on: sr404-self-hosted
+    runs-on: velox-self-hosted
```
Comment (jinchengchenghh) [inline .github/workflows/unittests.yml:30]: need fix

File: .github/workflows/unittests.yml [hunk 5]
```diff
@@ -27,38 +27,33 @@ concurrency:
...
       #   run: |
-      #     mvn test -Dcheckstyle.skip -Dexec.skip -Pbackends-gazelle -Dbuild_cpp=OFF -pl backends-gazelle -DwildcardSuites=io.glutenproject.execution.ArrowParquetWriteSuite -DfailIfNoTests=false
-      - name: Run TPCDS q82,q85,q37,q60,q26
-        run: |
-          cd /home/sparkuser/tpcds/
-          bash tpcds_spark.sh
+      #     mvn test -Dcheckstyle.skip -Dexec.skip -Pbackends-gazelle -pl backends-gazelle -DwildcardSuites=io.glutenproject.execution.ArrowParquetWriteSuite -DfailIfNoTests=false
+      #- name: Run TPCDS q82,q85,q37,q60,q26
+      #  run: |
+      #    docker exec velox-backend-test-$GITHUB_RUN_ID bash -c 'cd /opt/gluten && \
```
Comment (jinchengchenghh) [inline .github/workflows/unittests.yml:53]: Where is the test?

File: docs/GlutenUsage.md [hunk 6]
```diff
@@ -4,26 +4,9 @@ Based on the different environment, there are some parameters can be set via -D
...
 | backends-clickhouse                                   | Add -Pbackends-clickhouse in maven command to compile the JVM part of ClickHouse backend.                                                                                           | false                                              |
-| build_velox_backend                                   | Enable or Disable building the CPP part of Velox backend.                                                                                                                           | OFF                                                |
-| build_velox_from_source                               | Enable or Disable building Velox from a specific velox github repository. A default installed path will be in velox_home.                                                           | OFF                                                |
-| velox_home (only valid when _backend is ON)           | The path to the compiled Velox project. When building Gluten with Velox, if you have an existing Velox, please set it.                                                              | /PATH_TO_GLUTEN/ep/build-velox/build/velox_ep      |
```
Comment (jinchengchenghh) [inline docs/GlutenUsage.md:18]: maybe a mistake delete?

File: .github/workflows/unittests.yml (hunk 5 above)
Comment (zhejiangxiaomai) [reply .github/workflows/unittests.yml:53]: we will add later.

File: docs/Velox.md [hunk 7]
```diff
@@ -15,17 +16,54 @@ export JAVA_HOME=path/to/java/home
...
-The command below clones velox source code from [OAP-project/velox](https://github.com/oap-project/velox) to ep/build-velox/build/velox_ep. Then it applies some patches to Velox build script and builds the velox library.
+## install gcc and libraries to build arrow
+apt-get update && apt-get install -y sudo locales wget tar tzdata git ccache cmake ninja-build build-essential llvm-11-dev clang-11 libiberty-dev libdwarf-dev libre2-dev libz-dev libssl-dev libboost-all-dev libcurl4-openssl-dev openjdk-8-jdk maven
 
-```shell script
+## make sure jdk8 is used
+export JAVA_HOME=/usr/lib/jvm/java-8-openjdk-amd64
+export PATH=$JAVA_HOME/bin:$PATH
+
+#setup proxy on necessary
```
Comment (jinchengchenghh) [inline docs/Velox.md:30]: setup proxy if needed

File: .github/workflows/unittests.yml [hunk 8]
```diff
@@ -69,9 +64,23 @@ jobs:
           EXTRA_DOCKER_OPTIONS="--name velox-backend-test-$GITHUB_RUN_ID --detach" NON_INTERACTIVE=ON MOUNT_MAVEN_CACHE=OFF tools/gluten-te/cbash.sh sleep 14400
       - name: Install Gluten Velox-backend spark3.2 and Run Unit test
         run: |
-          docker exec velox-backend-test-$GITHUB_RUN_ID bash -c 'cd /opt/gluten && \
-          mvn clean install -Pspark-3.2 -Pspark-ut -Pbackends-velox -Dcpp_tests=ON -Dcpp_benchmarks=ON \
-            -Dbuild_velox_from_source=ON -Dbuild_arrow=ON'
+          docker exec velox-backend-test-$GITHUB_RUN_ID bash -c 'cd /opt/gluten/ep/build-arrow/src && \
+          ./get_arrow.sh  && \
+          sed -i 's/BUILD_TESTS=OFF/BUILD_TESTS=ON/' ./build_arrow_for_velox.sh && \
```
Comment (jinchengchenghh) [inline .github/workflows/unittests.yml:69]: Why add 'sed -i' here? Not change it in build_arrow_for_velox.sh

File: .github/workflows/unittests.yml [hunk 9]
```diff
@@ -27,38 +27,33 @@ concurrency:
...
-          rm -rf ~/.m2/repository/org/apache/arrow && rm -rf ~/.m2/repository/io/glutenproject/
-          mvn clean package -Pspark-3.2 -Pbackends-gazelle -DskipTests -Dbuild_cpp=ON -Dbuild_gazelle_cpp_backend=ON -Dbuild_arrow=ON
-          rm -rf /tmp/gluten_jars && mkdir -p /tmp/gluten_jars
-          cp package/gazelle/spark32/target/gluten-spark3.2_2.12-1.0.0-SNAPSHOT-jar-with-dependencies.jar /tmp/gluten_jars/
+          docker exec arrow-backend-test-$GITHUB_RUN_ID bash -c "cd /opt/gluten/ep/build-arrow/src && \
+          sed -i 's/backend_velox_main/arrow-8.0.0-gluten-20220427a/' ./get_arrow.sh && \
+          ./get_arrow.sh && \
+          ./build_arrow_for_gazelle.sh && \
+          cd /opt/gluten/cpp && \
+          sed -i 's/BUILD_GAZELLE_CPP_BACKEND=OFF/BUILD_GAZELLE_CPP_BACKEND=ON/' ./compile.sh && \
```
Comment (jinchengchenghh) [inline .github/workflows/unittests.yml:43]: It can use as an argument, not sed -i

File: tools/gluten-te/dockerfile-tpc [hunk 10]
```diff
@@ -12,9 +12,16 @@ RUN test -n "$BUILD_BACKEND_TYPE" || (echo "BUILD_BACKEND_TYPE not set" && false
...
+      && ./ep/build-velox/src/build_velox.sh 
     elif [ "$BUILD_BACKEND_TYPE" == "gazelle-cpp" ]; \
     then \
       GLUTEN_IT_MAVEN_OPTIONS="-Pspark-3.2 -Pgluten-gazellecpp -Darrow.version=8.0.0-gluten-SNAPSHOT"; \
+      && sed -i 's/backend_velox_main/arrow-8.0.0-gluten-20220427a/' ./ep/build-arrow/src/get_arrow.sh
+      && ./ep/build-arrow/src/get_arrow.sh \
+      && ./ep/build-arrow/src/build_arrow_for_gazelle.sh
     else \
       echo "Unrecognizable backend type: $BUILD_BACKEND_TYPE"; \
       exit 1; \
```
Comment (zhztheplayer) [inline tools/gluten-te/dockerfile-tpc:27]: The changes don't seem to work for  tpc docker. I'll help change this file before you merge

File: tools/gluten-te/dockerfile-buildenv [hunk 11]
```diff
@@ -64,4 +64,4 @@ RUN sed -i '/en_US.UTF-8/s/^# //g' /etc/locale.gen \
     && locale-gen
 ENV LANG en_US.UTF-8  
 ENV LANGUAGE en_US:en  
-ENV LC_ALL en_US.UTF-8
+ENV LC_ALL en_US.UTF-8
```
Comment (zhztheplayer) [inline tools/gluten-te/dockerfile-buildenv:67]: nit: preserve the white line as EoF

File: docs/GlutenUsage.md (hunk 6 above)
Comment (zhejiangxiaomai) [reply docs/GlutenUsage.md:18]: > maybe a mistake delete?

we add parameters to shell script and mvn will not use it.

File: .github/workflows/unittests.yml (hunk 8 above)
Comment (zhejiangxiaomai) [reply .github/workflows/unittests.yml]: fixed.

File: .github/workflows/unittests.yml (hunk 9 above)
Comment (zhejiangxiaomai) [reply .github/workflows/unittests.yml]: fixed.

File: tools/gluten-te/dockerfile-buildenv (hunk 11 above)
Comment (zhejiangxiaomai) [reply tools/gluten-te/dockerfile-buildenv:67]: Thanks for your suggestion.

File: docs/GlutenUsage.md (hunk 6 above)
Comment (jinchengchenghh) [reply docs/GlutenUsage.md:18]: Can you document shell script argument?

Comment (zhejiangxiaomai) [reply docs/GlutenUsage.md:18]: fixed. Added comments for all script argument.

**Review by zhztheplayer — COMMENTED** [review]
Is there any reason we don't provide a bundled build script anymore? If not I may suggest adding a shell script to build all stuffs just like the previous bundled mvn install script.

A bundled script can ease development specifically when enabling "ep cache" feature to automatically detect any updates from remote ep branch.

**Comment by zhztheplayer** [comment]
Is there any reason we don't provide a bundled build script anymore? If not I may suggest adding a shell script to build all stuffs just like the previous bundled mvn install script.

A bundled script can ease development specifically when enabling "ep cache" feature to automatically detect any updates from remote ep branch.

**Comment by zhztheplayer** [comment]
Like discussed, let's put off adding bundled build scripts.

File: docs/Velox.md [hunk 12]
```diff
@@ -15,17 +16,51 @@ export JAVA_HOME=path/to/java/home
...
+git clone https://github.com/oap-project/gluten.git
+
+
+## fetch arrow and compile
+cd /path_to_gluten/ep/build-arrow/src/
+./get_arrow.sh
+./build_arrow_for_velox.sh
+
+## fetch velox and compile
+cd /path_to_gluten/ep/build-velox/src/
```
Comment (zhixingheyi-tian) [inline docs/Velox.md:48]: for “## fetch velox and compile”,
is compiling in ./get_velox.sh or ./build_velox.sh?

Comment (zhejiangxiaomai) [reply docs/Velox.md:48]: ./get_velox.sh -> fetch velox
./build_velox.sh ->compile

**Review by zhouyuan — APPROVED** [review]
👍

**Comment by FelixYBW** [comment]
> Like discussed, let's put off adding bundled build scripts.

will this be added in another PR?

**Comment by zhejiangxiaomai** [comment]
> > Like discussed, let's put off adding bundled build scripts.
> 
> will this be added in another PR?
https://github.com/oap-project/gluten/pull/658 will add a bundled build scripts.

**Pull request merged**